### ISSUE TITLE
Some notes while following along Examples / 03 Locomotion

### ISSUE THREAD

**Issue by LarsJandel** [KybernetikGames/animancer#9]
- btw, thanks for the omit warning snippets, included in previous example scripts already: "#pragma warning disable CS0649 // Field is never assigned to, and will always have its default value." and similar 👍

03.01 Walk and Run
- nice follow-along tutorial 👌
- ah, here are you mentioning and linking to why using the "sealed" keyword I was missing in example 1, where it already occured
- "protected virtual void Reset()"
    - very handy feature! 💯
    - the "protected virtual void Reset()" part isn't present within the toggle script preview of "IdleAndWalk" at the top of the page
        - but it's included at the very bottom of the page
- probably v3 leftovers in the online documentation (provided scripts are fine):
    - "var otherState = Animancer.GetState(otherAnimation);" should be "var otherState = Animancer.States[otherAnimation];"
    - "_Animancer.CurrentState.Speed = movement;" should be "_Animancer.States.Current.Speed = movement;"
- thanks for the Cycle Offset info

03.02 Root Motion
- [Serializable] class: nice learning. Also how to define tooltips as I am just getting started with Unity
- maybe just my lack of C# knowledge: there are two "public void Play" methods: the two get merged later on / in the final script, but I found it confusing in the documentation, how can there be two methods with the same name, calling one from the other:
    public void Play(int index)
    {
        Play(_Animations[index]);// Call the Play method above.
    }
- switching to Transitions:
    - could help, if you would emphasize, that the previously set up MotionClip class gets replaced by the MotionTransition class, as this also updates the _Animations [SerializeField]
- you don't cover the use/concept of "Max Distance", which is part of the provided script, but yeah, this is not relevant to Animancer, but is a nice way to keep the character within the view. Thanks 
- Redirecting Root Motion: Quite flexible system - nice 👍

03.03 Linear Blending
- Transition Asset: I would prefer, if you would swap the order, explaining creating the Float 1 Controller Transition asset first and then go into the script as I got confused about the Float1ControllerTransition type until I finally got to that section. Thanks
- if you are extracting setting up UI buttons to a separate page, please consider including setting up the slider like in this example as well
    - and having a hint here to link the slider to the previously created public speed variables. As mentioned before, being a newbie, it took me a while to find this detail when setting up the scene from scratch. Thanks
- you could repeat how to create the controller transition asset with the Mixer to help memorizing this workflow

03.04 Directional Blending
- well, it is an intermediate example, I am not intermediate yet: I have to get into raycasting, vectors and angles more to better understand and memorize this example 😅
    - everything is commented, so that helps already
    - but fun to play with the Spider Bot, having it follow the mouse!

Thanks!

**Comment by KybernetikGames** [comment]
More great feedback.

> there are two "public void Play" methods

It's called method overloading (not to be confused with overriding an inherited method). You can have multiple methods with the same name as long as they each have different parameters. So since `_Animations[index]` is a `MotionClip`, using `Play(_Animations[index]);` will call the `Play(MotionClip)` method rather than `Play(int)`.

> Transition Asset: I would prefer, if you would swap the order, explaining creating the Float 1 Controller Transition asset first

That's tricky because I want to explain regular transitions and end on assets since that's what the script in the actual example uses. What if I just put a note in the [Blend Tree](https://kybernetikgames.github.io/animancer/v4-0/docs/examples/locomotion/linear-blending/#blend-tree) section that the script there is not the final one and will change below?

> well, it is an intermediate example, I am not intermediate yet

Let's fix that then :)

From what I can see, it could do with more explanation of raycasting, euler angles, and dot products. Is there anything else in particular you don't understand? Feel free to ask questions here too.

The rest of the examples haven't been updated for v4.0 yet (you can tell because their screenshots still have Unity's old GUI skin), but here's a quick summary of what I expect to change in case you want to go through them before I get to them:

- 04 Directional Sprites - pretty much nothing.
- 05 Animation Events - everything there is still relevant for using Animation Events, but I also need to demonstrate the new [Animancer Events](https://kybernetikgames.github.io/animancer/v4-0/docs/manual/events/) system and I'm not really sure how I'll go about that yet. Maybe I'll just have two characters in each scene and explain how to set up the exact same thing in both systems.
- 06 State Machines - not much new here either, though 06.05 Platformer now uses a Timeline Asset to play an introduction sequence when the scene starts which is shown in the [Change Log](https://kybernetikgames.github.io/animancer/v4-0/docs/changes/animancer-v4-0/#timeline).
- 07 Layers - CrossFade is now Play and the syntax for accessing layers changed a lot, but the general concepts are all still the same.
- 08 Inverse Kinematics - pretty much nothing.
- 09 Animator Controllers - not much in 01 Hybrid, though I want to try to make it walk over to the mini game position instead of teleporting. 02 3D Game Kit can have some things simplified by the new ability to specify custom end times and now that mixers have [Time Synchronisation](https://kybernetikgames.github.io/animancer/v4-0/docs/manual/blending/mixers/#synchronisation) they can replace some of the blend trees.

**Comment by LarsJandel** [comment]
Thanks for the info!

"Method overloading": Interesting - will keep that in mind, the possibility of using methods with the same name. Still wondering if this is a good practice or if there are even cases, where this has an advantage?

Transition order: Sorry, I think you misunderstood me. 😉 I was just talking about the actual "Transition Asset" paragraph. Hm well, thinking about it, the issue just happened as I was trying to follow-along line by line creating a scene from scratch, then I spotted the "Float1ControllerTransition", got confused, opened the provided example, saw the transition assets - the icons and looking at their inspector info - and didn't know, how they get created. Bottom line, you could add a line at the start of "Transition Asset", saying sth. like "If you are following along and create a scene from scratch, please read this paragraph to the end first." Yeah, the paragraph is not that long, maybe just my mistake not reading it to the end first... 😅

raycasting, euler angles, and dot products: Maybe you could add another screenshot where you draw some lines, what's happening. So direction is a vector pointing from one object to the other and direction.magnitude is a build-in method to receive the distance? Why do you use `< _MovementSpeed` to stop the bot? Couldn't this be 0 or any other minimum distance number?  The speed/distance relation and not using it sounds good, but got me confused as well: Doesn't the bot only has two speed options: normal and sprint, so that a bool value would be sufficient? Well, I guess I still have some trouble with direction/distance and of course the FixedUpdate `_Body.velocity = _MovementDirection * _Move.State.Speed * _MovementSpeed;`
Reading about dot products got me excited of course as this seems to be the future, so more info about that would be fantastic.
Btw, regarding the rotation: After copying that code and testing the scene I thought oh well, I do not see any active rotation. Reading the doc again: Ah okay, aligning to the camera! So, you should definitely motivate the reader/user to orbit the camera to see it in effect, plus maybe even adding another note to the UI info within the provided example scene to do so. You could change "click" to "mouse down / hold" for running while you are at it. Also, you should orbit the view in the gif within the documentation to showcase this behavior.

Thanks for the heads up regarding the remaining examples. I will have a look at them later.

**Comment by KybernetikGames** [comment]
There are lots of cases where method overloading is useful. For example, the [Play Methods](https://kybernetikgames.github.io/animancer/v4-0/docs/manual/playing/#play-methods) in Animancer all do very similar things with slight variations depending on how you want to identify the animation you want it to play and whether you want it to fade or not. Even the [Play Methods](https://docs.unity3d.com/ScriptReference/Animator.Play.html) in the regular Animator class has two overloads depending on how you want to identify the animation.

> So direction is a vector pointing from one object to the other and direction.magnitude is a build-in method to receive the distance?

Yes, magnitude is the length of the vector.

> Why do you use < _MovementSpeed to stop the bot? Couldn't this be 0 or any other minimum distance number?

It can't be 0 because it is extremely unlikely that moving from one frame to the next would end up at *exactly* the target location, so it needs to be a small threshold where we just stop if it gets close enough.

An arbitrary value like 0.1 or something might be too small if speed is too fast, so instead we can simply calculate the distance it will move in a single FixedUpdate (`_MovementSpeed * Time.fixedDeltaTime`) and use that as the threshold. Though now that you mention it, the `_Move.State.Speed` should also be included there since it is part of the velocity.

> Also, you should orbit the view in the gif within the documentation to showcase this behavior.

I actually did that the first time I recorded it, but I screwed up the movements a bit and then got distracted and forgot about it by the time I went to re-record it :)
